### ISSUE TITLE
Slow startup issues

### ISSUE THREAD

**Issue by One1One1One** [Bionus/imgbrd-grabber#1397]
**Bug description**

The startup time of this program is getting slower over time as the main.log (and possibly the `commands.sql`) fills up. In my case a `main.log` of 123MB and a `commands.sql` of 90MB caused my Raspberry Pi to take 10 minutes for this program to start while one of its cores used 100% CPU compared to less than 1 second of startup time when both of the files are empty.

**Steps to reproduce**

1. (optional) for the log to fill up quickly, input your database settings
2. Download a lot of files
3. Close the program if it didn't already crash because it used too much RAM
4. restart and wait for ~10 minutes

**Expected behavior**

A fast startup like if both of these files were empty.

**Opinion**

I'm not too familiar with Qt and multiplatform programming but from regular C++ programming and other languages I know that getting a file handle for write access to a file hardly requires any memory and CPU. I don't understand why grabber even creates `commands.sql`. I checked my database and all of the commands inside this file were already executed.

**System information**

- OS: Raspbian
- Grabber version: #afa4403 (newest commit from dev)

**Comment by One1One1One** [comment]
I just checked the log tab in grabber. It's literally loading the whole log file into the GUI. That explains a lot. I thought it is working similar to to command `tail -n 20` on linux so it only loads the last 20 lines on startup and deletes the first lines upon adding new lines when the current number of lines exceed 20.

How about implementing a log level filter, so it only logs warning or above or only errors?

How about limiting the number of lines it loads on startup?

How about limiting the number of lines in error log tab? If it exceeds this number when a new line is added the oldest line will get deleted.

How about purging the log file when the program is closed normally?

How about making the `commands.sql` optional? At least for me it serves no purpose.

**Comment by Bionus** [comment]
Both of those files are opened in truncate mode at startup ([logger.cpp](https://github.com/Bionus/imgbrd-grabber/blob/master/lib/src/logger.cpp#L15)). On my Windows machine, these logs are properly cleared every time the program starts.

I just noticed however that the flags provided don't really make sense (opened in both append and truncate mode), so maybe there's a different flag priority on Windows and Rasbian, where on Windows the file would properly be truncated, but be opened in append mode on Rasbian.

This decision to clear logs on startup is questionable, but was done for simplicity and to prevent overgrowing logs (like in your situation), or log rotation (which is a pain and a waste of disk space).

> How about implementing a log level filter, so it only logs warning or above or only errors? That could help keeping the log size small while making it easier to check the file for errors on a regular basis.

There's already such filters (via `Logger::setLogLevel`), set on `Info` by default and can be set to `Debug` via the `-d` CLI flag. Do you mean to be able to set it to other levels (in your case `Warning) via a CLI option?

> How about limiting the number of lines it loads on startup?

Well after they are truncated properly, this shouldn't be an issue anymore. The only reason the log is read on startup is to show log calls between the time the process started and the main window opened.

> How about limiting the number of lines in error log tab? If it exceeds this number when a new line is added the oldest line will get deleted.

That's a possibility indeed, not sure whether there would be any visible gain however.

> How about purging the log file when the program is closed normally?

Sounds redundant to a startup clear?

>I don't understand why grabber even creates commands.sql. I checked my database and all of the commands inside this file were already executed.

`commands.sql` is, just like `main.log`, simply a log file. Except this one logs every single SQL command executed by the program.

**Comment by Bionus** [comment]
Does `strace -eopen -- ./Grabber` not show something like:

```
open("main.log", O_WRONLY | O_CREAT | O_APPEND | O_TRUNC, 0666) = 3
```

?

**Comment by One1One1One** [comment]
It shows:
`open("/mnt/drive/grabber/main.log", O_RDONLY|O_LARGEFILE|O_CLOEXEC) = 12`

**Comment by BarryMode** [comment]
I noticed that the program has been starting up slowly in the last several versions. Sure enough, my main.log was 105mb and it seemed to be starting slower and slower. As soon as I deleted that file, it started up instantly. As usual, on Mac.

Also as a completely unrelated note, I thought I'd share that the program works with MacOS Mojave dark mode. Thought you might want to see it.

![](https://user-images.githubusercontent.com/5648875/46318289-9aa66900-c59b-11e8-9db8-1e40ea5f2a43.png)
